third_party/txt/src/txt/paragraph.cc

```diff
@@ -1394,6 +1386,48 @@ void Paragraph::PaintDecorations(SkCanvas* canvas,
   }
 }
 
+void Paragraph::ComputeWavyDecoration(SkPath& path,
+                                      double x,
+                                      double y,
+                                      double width,
+                                      double thickness) {
+  int wave_count = 0;
+  double x_start = 0;
+  // One full wavelength is 4 * thickness.
+  double quarter = thickness;
+  path.moveTo(x, y);
+  double remaining = width;
+  while (x_start + quarter * 2 < width) {
+    path.rQuadTo(quarter, wave_count % 2 == 0 ? -quarter : quarter, quarter * 2,
+                 0);
+    remaining = width - (x_start + quarter * 2);
+    x_start += quarter * 2;
+    ++wave_count;
+  }
+  // Manually add a final partial quad for the remaining width that do
+  // not fit nicely into a half-wavelength.
+  // The following math is based off of quadratic bezier equations:
+  //
+  //  * Let P(x) be the equation for the curve.
+  //  * Let P0 = start, P1 = control point, P2 = end
+  //  * P(x) = -2x^2 - 2x
+  //  * P0 = (0, 0)
+  //  * P1 = 2P(0.5) - 0.5 * P0 - 0.5 * P2
+  //  * P2 = P(remaining / (wavelength / 2))
+  double normalized_remaining = remaining / (quarter * 2);
```

Comment:
Using the standard subdivision method for quads, would involve the following steps:

```
// x,y values for full quad:
// This math is for the y = +quarter piece,
// but you can just negate y1,y2 at the end if it is the other piece
x0 = 0;
y0 = 0;
x1 = quarter;
y1 = quarter;
x2 = quarter * 2;
y2 = 0;
// now compute the subdivision at t (== normalized_remaining)
// xy2 = xy1 + t * (xy2 - xy1);
x2 = x1 + t * (x2 - x1);
    = quarter + t * (2quarter - quarter);
    = quarter + t*quarter;
    = (1 + t) * quarter;
y2 = y1 + t * (y2 - y1);
   = quarter + t * (0 - quarter);
   = (1 - t) * quarter;
// xy1 = xy0 + t * (xy1 - xy0);
x1 = x0 + t * (x1 - x0);
   = 0 + t * (quarter - 0);
   = t * quarter;
y1 = y0 + t * (y1 - y0);
   = 0 + t * (quarter - 0);
   = t * quarter;
// xy2 = xy1 + t * (xy2 - xy1);
x2 = x1 + t * (x2 - x1);
   = t * quarter + t * ((1 + t) * quarter - t * quarter);
   = t * quarter + t * (quarter);
   = 2t * quarter;
y2 = y1 + t * (y2 - y1);
   = t * quarter + t * ((1 - t) * quarter - t * quarter);
   = t * quarter + (1 - 2t) * t * quarter;
   = (2 - 2t) * t * quarter;
// Final equations are:
x1 = t * quarter;
y1 = t * quarter;
x2 = 2t * quarter;
y2 = (2 - 2t) * t * quarter;
// Since we have t == normalized_remaining == remaining / (quarter * 2),
// Then t * quarter == remaining / 2;
// Substituting for t:
x1 = remaining / 2;
y1 = remaining / 2;
x2 = remaining;
y2 = (2 - remaining/quarter) * remaining / 2;
   = remaining - remaining*remaining/(quarter*2);
// then negate y1,y2 if wave_count is odd...
```

Reply:
I'm not sure why, but it looks like my equation for y2 is negated from yours...?

Reply:
(for reference, unsimplified version)
```
  double normalized_remaining = remaining / (quarter * 2);
  double p2_x = remaining;
  double p2_y = (wave_count % 2 != 0 ? -1 : 1) * quarter *
                (2 * normalized_remaining * normalized_remaining -
                 2 * normalized_remaining);
  double mid_x = remaining / 2;
  double mid_y =
      (wave_count % 2 != 0 ? -1 : 1) * quarter *
      (normalized_remaining * normalized_remaining / 2 - normalized_remaining);
  double p1_x = 2 * mid_x - p2_x / 2;
  double p1_y = 2 * mid_y - p2_y / 2;
  path.rQuadTo(p1_x, p1_y, p2_x, p2_y);
```